docs/adr/0005-feature-flags.md

```diff
@@ -0,0 +1,149 @@
+
+When applying feature flagging, we often have a choice between making that toggle on the frontend or the backend. Here are a few factors that we should keep in mind while making that decision
+
+### UI Performance
+By moving flagging decisions to the server side, we are reducing code bloat on the frontend, gaining performance that's immediately noticeable by the user. If we decide to move Monitor frontend to SPA in the future, Single-page applications are already making a API call to render the data needed for the UI. With the same payload, we can make a call to a feature-flag service, so one network call can contain all feature-flag configs with the server-side data.
+
+### Config Cache Lag 
+There are two general approaches:
+
+1. An app can proactively request flagging decisions based on runtime-context
+
+> This approach consumes a lot more bandwidth (every context changes = new feature flag decision being made by the backend). It's a lot more flexible and prevents cache lag
+2. An app can request general config and use client-side router
+
+> With this approach, we essentially cache the config. In the case of a kill switch, the frontend may not be able to react quickly to the changes from the backend.
+
+### Security
+One key consideration when deciding where to implement a feature flag is security. If a feature is highly sensitive or critical to the security of the application, it may be safer to implement the toggle on the backend. This can help ensure that the feature is properly secured and that user data is protected.
+
+### Sync Complexity
+Implementing feature flagging on the server side reduces the complexity of synchronizing feature toggle logic between the frontend and backend, as there is a single central location to manage the toggle settings. This simplifies development and ensures that the system runs smoothly, as dependencies flow in one direction.
+## Maintenance & Best Practices
+### Naming Flags
+Defining a pattern or naming convention is generally a good practice for keeping things organized. Engineers can quickly identify from the name what the feature is about and be able to recognize what areas of the application for which the feature is being used. I'll list a few useful examples of naming structures and demonstrate the practical benefits.
+
+A good naming **convention** I found is as follows:
+```
+section_purpose_service
+```
+
+The feature name example that follows has three parts:
+1. section (optional): we present the name of the section of the app the feature is gating (free/premium). 
+2. purpose: indicates what the feature does (can be tied to an epic name)
+3. service: where in the stack the feature is located (subsection of an epic)
+
+An **example** of such names would be:
+```
+premium_subplat_email
+```
+
+In this example, the feature is gating functionality in the "settings" section. It is a part of the "premium" feature. The main area of concern is the "email"
+service.
+
+### Retiring Flags / Zombies
```

Comment:
Yes! Thank you for including this :clap: One other radical suggestion I've seen is to disallow feature flags enabling something for everyone, i.e. you can roll out to 95% of users, but to roll out to everyone, you need to remove the flag — but that's probably a bit too radical. :sweat_smile:

One thing that's been helpful on Relay is to define a union of strings [enumerating all active feature flags](https://github.com/mozilla/fx-private-relay/blob/57dde8ff1c5794330232e5ee899870e79facbab1/frontend/src/hooks/api/runtimeData.ts#L6-L16) in TypeScript. Then we can just remove a flag from that type, and TS will tell us which bits of code will need to be updated - because it's often not entire if/else branches you can just delete.